File: android/src/main/kotlin/so/engage/engage_flutter/MethodCallHandler.kt
package so.engage.engage_flutter

import android.content.Context
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal class MethodCallHandler(
    context: Context,
    eventChannel: EventChannel,
    methodChannel: MethodChannel
): MethodChannel.MethodCallHandler {
    private var context: Context
    private var eventChannel: EventChannel
    private var methodChannel: MethodChannel

    init {
        this.context = context
        this.eventChannel = eventChannel
        this.methodChannel = methodChannel
    }

//    private lateinit var engage: Engage

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when(call.method) {
            "initialise" -> initialise(call, result)
            "identify" -> identify(call, result)
            "setDeviceToken" -> setDeviceToken(call, result)
            "logout" -> logout(call, result)
            "addToAccount" -> addToAccount(call, result)
            "addAttributes" -> addAttributes(call, result)
            "removeFromAccount" -> removeFromAccount(call, result)
            "changeAccountRole" -> changeAccountRole(call, result)
            "convertToCustomer" -> convertToCustomer(call, result)
            "convertToAccount" -> convertToAccount(call, result)
            "merge" -> merge(call, result)
            "track" -> track(call, result)
            else -> result.notImplemented()
        }
    }

    private fun initialise(call: MethodCall, result: MethodChannel.Result) {
        val publicKey: String? = call.argument("publicKey")
//        engage = Engage(context, publicKey)
    }

    private fun identify(call: MethodCall, result: MethodChannel.Result) {
        val uid: String = call.argument("uid") ?: return
        val properties: Map<String, Any> = call.argument("properties") ?: return
        // engage.idenify
    }

    private fun setDeviceToken(call: MethodCall, result: MethodChannel.Result) {
        val deviceToken: String = call.argument("deviceToken") ?: return
        val uid: String? = call.argument("uid")
        // engage.setDeviceToken
    }

    private fun logout(call: MethodCall, result: MethodChannel.Result) {
        val deviceToken: String = call.argument("deviceToken") ?: return
        val uid: String? = call.argument("uid")
        // engage.logout
    }

    private fun addToAccount(call: MethodCall, result: MethodChannel.Result) {
        val aid: String = call.argument("aid") ?: return
        val role: String? = call.argument("role")
        val uid: String? = call.argument("uid")
        // engage.addToAccount
    }

    private fun addAttributes(call: MethodCall, result: MethodChannel.Result) {
        val properties: Map<String, Any> = call.argument("properties") ?: return
        val uid: String? = call.argument("uid")
        // engage.addAttributes
    }

    private fun removeFromAccount(call: MethodCall, result: MethodChannel.Result) {
        val aid: String = call.argument("aid") ?: return
        val uid: String? = call.argument("uid")
        // engage.removeFromAccount
    }

    private fun changeAccountRole(call: MethodCall, result: MethodChannel.Result) {
        val aid: String = call.argument("aid") ?: return
        val role: String = call.argument("role") ?: return
        val uid: String? = call.argument("uid")
        // engage.changeAccountRole
    }

    private fun convertToCustomer(call: MethodCall, result: MethodChannel.Result) {
        val uid: String? = call.argument("uid")
        // engage.convertToCustomer
    }

    private fun convertToAccount(call: MethodCall, result: MethodChannel.Result) {
        val uid: String? = call.argument("uid")
        // engage.convertToAccount
    }

    private fun merge(call: MethodCall, result: MethodChannel.Result) {
        val source: String = call.argument("source") ?: return
        val destination: String = call.argument("destination") ?: return
        val uid: String? = call.argument("uid")
        // engage.changeAccountRole
    }

    private fun track(call: MethodCall, result: MethodChannel.Result) {
        val event: String = call.argument("event") ?: return
        val properties: Map<String, Any> = call.argument("properties") ?: return
        val uid: String? = call.argument("uid")
        // engage.track
    }
}

File: ios/Classes/MethodCallHandler.swift
import Flutter
import UIKit
import Engage

public class MethodCallHandler: NSObject {
    var eventChannel: FlutterEventChannel
    var methodChannel: FlutterMethodChannel
    
    init(eventChannel: FlutterEventChannel, methodChannel: FlutterMethodChannel) {
        self.eventChannel = eventChannel
        self.methodChannel = methodChannel
        super.init()
    }
    
    let engage: Engage = Engage.shared
    
    public func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialise":
            initialise(call, result: result)
            return
        case "identify":
            identify(call, result: result)
            return
        case "setDeviceToken":
            setDeviceToken(call, result: result)
            return
        case "logout":
            logout(call, result: result)
            return
        case "addToAccount":
            addToAccount(call, result: result)
            return
        case "addAttributes":
            addAttributes(call, result: result)
            return
        case "removeFromAccount":
            removeFromAccount(call, result: result)
            return
        case "changeAccountRole":
            changeAccountRole(call, result: result)
            return
        case "convertToCustomer":
            convertToCustomer(call, result: result)
            return
        case "convertToAccount":
            convertToAccount(call, result: result)
            return
        case "merge":
            merge(call, result: result)
            return
        case "track":
            track(call, result: result)
            return
        default:
            result(FlutterMethodNotImplemented)
            return
        }
    }
    
    
    private func initialise(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>,
           let publicKey = args["publicKey"] as? String {
            let _ = engage.initialise(publicKey: publicKey)
        }
    }
    
    private func identify(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>,
           let uid = args["uid"] as? String,
           let properties = args["properties"] as? Dictionary<String, Any> {
            engage.identify(uid: uid, properties: properties)
        }
    }
    
    private func setDeviceToken(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>,
           let deviceToken = args["deviceToken"] as? String {
            let uid = args["uid"] as? String
            engage.setDeviceToken(deviceToken: deviceToken, uid: uid)
        }
    }
    
    private func logout(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>,
           let deviceToken = args["deviceToken"] as? String {
            let uid = args["uid"] as? String
            engage.logout(deviceToken: deviceToken, uid: uid)
        }
    }
    
    private func addToAccount(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>,
           let aid = args["aid"] as? String {
            let role = args["role"] as? String
            let uid = args["uid"] as? String
            engage.addToAccount(aid: aid, role: role, uid: uid)
        }
    }
    
    private func addAttributes(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>,
           let properties = args["properties"] as? Dictionary<String, Any> {
            let uid = args["uid"] as? String
            engage.addAttributes(properties: properties, uid: uid)
        }
    }
    
    private func removeFromAccount(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>,
           let aid = args["aid"] as? String {
            let uid = args["uid"] as? String
            engage.removeFromAccount(aid: aid, uid: uid)
        }
    }
    
    private func changeAccountRole(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>,
           let aid = args["aid"] as? String,
           let role = args["role"] as? String {
            let uid = args["uid"] as? String
            engage.changeAccountRole(aid: aid, role: role, uid: uid)
        }
    }
    
    private func convertToCustomer(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any> {
            let uid = args["uid"] as? String
            engage.convertToCustomer(uid: uid)
        }
    }
    
    private func convertToAccount(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any> {
            let uid = args["uid"] as? String
            engage.convertToAccount(uid: uid)
        }
    }
    
    private func merge(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>,
           let source = args["source"] as? String,
           let destination = args["destination"] as? String {
            engage.merge(source: source, destination: destination)
        }
    }
    
    private func track(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? Dictionary<String, Any>,
           let event = args["event"] as? String {
            let value = args["value"] as? Dictionary<String, Any>
            let date = args["date"] as? Date
            let uid = args["uid"] as? String
            engage.track(event: event, value: value, date: date, uid: uid)
        }
    }
}
